### PR TITLE
Add some new block events and dispenser helper methods

### PR DESCRIPTION
**Pull request by abigailfails**
As per the conversation in DMs, `AnimateTickEvent`, `EntityWalkEvent` and `AnimateFluidTickEvent`  allow you to make changes to the behavior defined in the respective methods in `Block` and `Fluid`. Cancelling these prevents the original method from being called, so you could, for example, replace the particle the enchantment tables give off.

The dispenser helper methods in `BlockUtil` are just there because I found out that it was really tedious to write statements to find the block or entities in front of the dispenser in the predicate lambda for `registerAlternativeDispenseBehavior`. It may be the wrong class to put it in and the javadocs definitely need reviewing, I wasn't sure how to space out @param and @return properly.

### CODE REVIEW

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java [hunk 1]
```diff
@@ -118,4 +121,42 @@ public static BBRotation getRotationForDirection(Direction currentDirection, Dir
 			}
 		}
 	}
+
+	/**
+	 * <p>Gets the {@link BlockPos} in front of a dispenser.</p>
+	 * {@link BlockState#getBlock() source.getBlockState().getBlock()} must return an instance of {@link DispenserBlock}.
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:127]: This is incorrect, the state must have the `DirectionalBlock.FACING` tag.

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java [hunk 2]
```diff
@@ -118,4 +121,42 @@ public static BBRotation getRotationForDirection(Direction currentDirection, Dir
...
+	 *
+	 * @param source The {@link IBlockSource} to get the position from.
+	 * @return The position in front of the dispenser's output face.
+	 *
+	 * @author abigailfails
+	 * */
+	public static BlockPos dispenserOffsetPos(IBlockSource source) {
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:134]: I don't like this method being called `dispenserOffsetPos()` cause that is not what it does exactly.
In actuality, it just offsets the position in the direction of a state's `DirectionalBlock.FACING` property.

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java [hunk 3]
```diff
@@ -118,4 +121,42 @@ public static BBRotation getRotationForDirection(Direction currentDirection, Dir
...
+		return source.getBlockPos().offset(source.getBlockState().get(DispenserBlock.FACING));
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:135]:
```suggestion
		return source.getBlockPos().offset(source.getBlockState().get(DirectionalBlock.FACING));
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java [hunk 4]
```diff
@@ -118,4 +121,42 @@ public static BBRotation getRotationForDirection(Direction currentDirection, Dir
...
+	}
+
+	/**
+	 * Gets the {@link BlockState} at the position in front of a dispenser.
+	 *
+	 * @param source The {@link IBlockSource} to get the position from.
+	 * @return The {@link BlockState} at the position in front of the dispenser's output face.
+	 * @see #dispenserOffsetPos(IBlockSource source)
+	 * */
+	public static BlockState stateAtDispenserOffsetPos(IBlockSource source) {
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:145]:
```suggestion
	public static BlockState getStateAtDispenserOffset(IBlockSource source) {
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java [hunk 5]
```diff
@@ -118,4 +121,42 @@ public static BBRotation getRotationForDirection(Direction currentDirection, Dir
...
+		return source.getWorld().getBlockState(dispenserOffsetPos(source));
+	}
+
+	/**
+	 * Gets the {@link List}<{@link Entity}> of entities at the position in front of a dispenser.
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:150]:
```suggestion
	 * Gets a {@link List} of entities at the position in front of a dispenser.
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java [hunk 6]
```diff
@@ -118,4 +121,42 @@ public static BBRotation getRotationForDirection(Direction currentDirection, Dir
...
+	 *
+	 * @param source The {@link IBlockSource} to get the position from.
+	 * @param entityType The class extending {@link Entity} to search for. Set to {@code Entity.class} to get all
+	 *                   entities, regardless of type.
+	 *
+	 * @return The {@link List}<{@link Entity}> of entities at the position in front of the dispenser's output face.
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:156]:
```suggestion
	 * @return A {@link List} of entities at the position in front of the dispenser's output face.
```

**Review by SmellyModder — CHANGES_REQUESTED** [review]

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java (hunk 2 above)
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:135]: Oh yeah, forgot that `DispenserBlock.FACING` is just `DirectionalBlock.FACING`. I'll update the names and documentation to reflect that

File: src/main/java/com/minecraftabnormals/abnormals_core/core/events/AnimateTickEvent.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    public AnimateTickEvent(BlockState state, World world, BlockPos pos, Random rand) {
+        super(world, pos, state);
+        this.random = rand;
+    }
+
+    public Random getRandom() {
+        return random;
+    }
+
+    public static boolean onAnimateTick(BlockState state, World world, BlockPos pos, Random rand) {
```
[the remark this answers is not in the document]
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/events/AnimateTickEvent.java:40]: Added in latest commit

File: src/main/java/com/minecraftabnormals/abnormals_core/core/events/AnimateFluidTickEvent.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    public FluidState getState() {
+        return state;
+    }
+
+    public Random getRandom() {
+        return random;
+    }
+
+    public static boolean onAnimateFluidTick(World world, BlockPos pos, FluidState state, Random random) {
```
[the remark this answers is not in the document]
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/events/AnimateFluidTickEvent.java]: Added in latest commit

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java (hunk 2 above)
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java]: Refactoring done in latest commit.

**Comment by abigailfails** [comment]
I think I've addressed all your requested changes now.

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java (hunk 1 above)
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java]: Updated in latest commit

File: src/main/resources/abnormals_core.mixins.json [hunk 9]
```diff
@@ -12,9 +12,11 @@
     "ServerWorldMixin",
     "SpawnEggItemMixin",
     "TrackedEntityMixin",
-    "TrapDoorBlockMixin"
+    "TrapDoorBlockMixin",
+    "IFluidMixin"
   ],
   "client": [
+    "client.ClientWorldMixin",
```
[the remark this answers is not in the document]
Comment (abigailfails) [reply src/main/resources/abnormals_core.mixins.json]: Fixed in latest commit

File: src/main/java/com/minecraftabnormals/abnormals_core/core/events/AnimateTickEvent.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        this.random = rand;
+    }
+
+    public Random getRandom() {
+        return this.random;
+    }
+
+    /**
+     * Fires the {@link AnimateTickEvent} for a given {@link BlockState}, {@link World}, {@link BlockPos} and {@link Random}.
+     * @param state  The {@link BlockState} that {@link net.minecraft.block.Block#animateTick Block.animateTick()} is being fired for.
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/events/AnimateTickEvent.java:35]:
```suggestion
     * @param state The {@link BlockState} that {@link net.minecraft.block.Block#animateTick Block.animateTick()} is being fired for.
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java [hunk 11]
```diff
@@ -118,4 +121,46 @@ public static BBRotation getRotationForDirection(Direction currentDirection, Dir
...
+	public static BlockState getStateAtOffsetPos(IBlockSource source) {
+		return source.getWorld().getBlockState(offsetPos(source));
+	}
+
+	/**
+	 * Gets a {@link List} of entities at the position returned by {@link #offsetPos(IBlockSource source)}.
+	 *
+	 * @param source The {@link IBlockSource} to get the position from.
+	 * @param entityType The class extending {@link Entity} to search for. Set to {@code Entity.class} to get all
+	 *                   entities, regardless of type.
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:157]:
```suggestion
	 * @param entityType The class extending {@link Entity} to search for. Set to {@code Entity.class} to get all entities, regardless of type.
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java [hunk 12]
```diff
@@ -118,4 +121,46 @@ public static BBRotation getRotationForDirection(Direction currentDirection, Dir
...
+		return source.getBlockPos().offset(source.getBlockState().get(DirectionalBlock.FACING));
+	}
+
+	/**
+	 * Gets the {@link BlockState} at the position returned by {@link #offsetPos(IBlockSource source)}.
+	 *
+	 * @param source The {@link IBlockSource} to get the position from.
+	 * @return The {@link BlockState} at the offset position.
+	 * @see #offsetPos(IBlockSource source)
+	 *
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:146]:
```suggestion
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java [hunk 13]
```diff
@@ -118,4 +121,46 @@ public static BBRotation getRotationForDirection(Direction currentDirection, Dir
...
+	/**
+	 * <p>Returns the {@link BlockPos} offset by 1 in the direction of {@code source}'s {@link BlockState}'s
+	 * {@link DirectionalBlock#FACING} property.</p>
+	 * This requires the {@link BlockState} stored in {@code source} to have a {@link DirectionalBlock#FACING} property.
+	 *
+	 * @param source The {@link IBlockSource} to get the position from.
+	 * @return The position in front of the dispenser's output face.
+	 *
+	 * @author abigailfails
+	 *
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:134]:
```suggestion
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java [hunk 14]
```diff
@@ -118,4 +121,46 @@ public static BBRotation getRotationForDirection(Direction currentDirection, Dir
...
+	/**
+	 * Gets a {@link List} of entities at the position returned by {@link #offsetPos(IBlockSource source)}.
+	 *
+	 * @param source The {@link IBlockSource} to get the position from.
+	 * @param entityType The class extending {@link Entity} to search for. Set to {@code Entity.class} to get all
+	 *                   entities, regardless of type.
+	 *
+	 * @return A {@link List} of entities at the at the offset position.
+	 * @see #offsetPos(IBlockSource source)
+	 *
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/BlockUtil.java:161]:
```suggestion
```

**Review by SmellyModder — CHANGES_REQUESTED** [review]
Looking better now.
A few more nitpicks.

File: src/main/java/com/minecraftabnormals/abnormals_core/core/mixin/IFluidMixin.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+import net.minecraft.fluid.FluidState;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Invoker;
+
+import java.util.Random;
+
+@Mixin(Fluid.class)
+public interface IFluidMixin {
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/mixin/IFluidMixin.java:13]:
```suggestion
public interface FluidAccessorMixin {
```

**Review by SmellyModder — CHANGES_REQUESTED** [review]

**Comment by abigailfails** [comment]
Is this going to have trouble merging with the mixins json?

**Review by SmellyModder — APPROVED** [review]

**Pull request merged**